File: tests/test_plan_reuse_invalidation.py
#!/usr/bin/env python3
"""Copied-fixture invalidation checks for selective plan reuse."""
import argparse
import hashlib
import json
import os
from pathlib import Path
import shutil
import subprocess
import tempfile


ROOT = Path(__file__).resolve().parents[1]
UNSUPPORTED_ENV = ("LD_PRELOAD", "LD_LIBRARY_PATH", "DYLD_INSERT_LIBRARIES",
                   "DYLD_LIBRARY_PATH", "ENV", "BASH_ENV")


def digest(value):
    return hashlib.sha256(value).hexdigest()


def json_digest(value):
    return digest(json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":")).replace("/", r"\/").encode())


def find_fixture(fixture):
    homes = fixture / "home"
    runs = list(homes.glob("state/v2/projects/*/workflows/runs/*"))
    if len(runs) != 1:
        raise AssertionError(f"expected one workflow run under {homes}, found {runs}")
    run = runs[0]
    source = fixture / "source"
    if not source.is_dir():
        raise AssertionError(f"missing fixture source: {source}")
    return homes, run.name, source


def recorded_environment(run, home, fleet_bin):
    manifest = json.loads((run / "artifacts" / "environment").read_text())
    variables = manifest["variables"]
    env = dict(os.environ)
    env.update({"HYDRA_HOME": str(home), "HYDRA_FLEET_BIN": str(fleet_bin)})
    env.update(variables)
    for name in UNSUPPORTED_ENV:
        env.pop(name, None)
    return env


def run_result(source, home, run_id, fleet_bin, expect_pass):
    env = recorded_environment(run_path(home, run_id), home, fleet_bin)
    command = [str(ROOT / "bin" / "hydra"), "workflow", "plan", "result", run_id]
    result = subprocess.run(command, cwd=source, env=env, capture_output=True,
                            text=True, timeout=45)
    if expect_pass:
        if result.returncode or not json.loads(result.stdout)["ok"]:
            raise AssertionError(result.stdout + result.stderr)
    elif result.returncode == 0:
        raise AssertionError("corrupted copied fixture was accepted: " + result.stdout)
    return result


def run_path(home, run_id):
    runs = list(home.glob("state/v2/projects/*/workflows/runs"))
    if len(runs) != 1:
        raise AssertionError(f"expected one copied project, found {runs}")
    return runs[0] / run_id


def rehash_reuse_record(run, step, file_name):
    record_path = run / "repair-2.json"
    record = json.loads(record_path.read_text())
    proof = record["evidence"]["reuse"]["steps"][step]
    proof[file_name] = digest((run / "steps" / step / "attempt-1" / file_name).read_bytes())
    record["sha256"] = json_digest(record["evidence"])
    record_path.write_text(json.dumps(record, sort_keys=True, separators=(",", ":")) + "\n")


def tamper_input(run):
    path = run / "steps/produce/attempt-1/inputs/environment"
    path.write_bytes(path.read_bytes() + b" ")


def tamper_environment(run):
    path = run / "steps/produce/attempt-1/inputs/environment"
    value = json.loads(path.read_text())
    value["scope"] = "tampered"
    path.write_text(json.dumps(value, sort_keys=True) + "\n")


def tamper_recipe(run):
    path = run / "steps/produce/attempt-1/inputs/recipe"
    path.write_bytes(path.read_bytes() + b" ")


def tamper_acceptance(run):
    (run / "plan-accepted").write_text("0" * 64 + "\n")


def tamper_receipt(run):
    path = run / "steps/produce/attempt-1/remote/receipt.json"
    value = json.loads(path.read_text())
    value["submission_key"] = "tampered"
    path.write_text(json.dumps(value, sort_keys=True) + "\n")
    rehash_reuse_record(run, "produce", "remote/receipt.json")


def tamper_result(run):
    path = run / "steps/produce/attempt-1/remote/result.json"
    value = json.loads(path.read_text())
    value["result"]["receipt"]["task_id"] = "task_" + "0" * 64
    path.write_text(json.dumps(value, sort_keys=True) + "\n")
    rehash_reuse_record(run, "produce", "remote/result.json")


def tamper_failures(run):
    path = run / "repair-2.json"
    value = json.loads(path.read_text())
    value["evidence"]["failures"] = []
    value["sha256"] = json_digest(value["evidence"])
    path.write_text(json.dumps(value, sort_keys=True, separators=(",", ":")) + "\n")


def tamper_compiled(run):
    path = run / "compiled.json"
    value = json.loads(path.read_text())
    value["plan"]["objective"] += " tampered"
    path.write_text(json.dumps(value, sort_keys=True) + "\n")


def tamper_source_binding(run):
    path = run / "compiled.json"
    value = json.loads(path.read_text())
    value["source"]["sha256"] = "0" * 64
    path.write_text(json.dumps(value, sort_keys=True) + "\n")


def tamper_unknown_step(run):
    path = run / "repair-2.json"
    value = json.loads(path.read_text())
    value["evidence"]["reuse"]["steps"]["unknown"] = {}
    value["sha256"] = json_digest(value["evidence"])
    path.write_text(json.dumps(value, sort_keys=True, separators=(",", ":")) + "\n")


def tamper_dependency_removal(run):
    path = run / "repair-2.json"
    value = json.loads(path.read_text())
    del value["evidence"]["reuse"]["steps"]["inspect"]
    value["sha256"] = json_digest(value["evidence"])
    path.write_text(json.dumps(value, sort_keys=True, separators=(",", ":")) + "\n")


def policy_controls(fixture, source, home, run_id, fleet_bin):
    original = json.loads((fixture / "plan.json").read_text())
    env = recorded_environment(run_path(home, run_id), home, fleet_bin)
    changes = [
        ("policy-version", ("reuse_policy", "schema_version"), 2),
        ("policy-mode", ("reuse_policy", "mode"), "external_cache"),
        ("policy-extra-field", ("reuse_policy", "extra"), True),
        ("unknown-step", ("reuse_policy", "steps", "unknown"), {}),
        ("incomplete-dependencies", ("reuse_policy", "steps", "produce", "dependencies"), "unknown"),
        ("external-effects", ("reuse_policy", "steps", "produce", "effects"), "external"),
        ("unbound-environment", ("reuse_policy", "steps", "produce", "environment_input"), "missing"),
        ("repair-input", ("data", "steps", "produce", "inputs", "repair"), {"repair": True}),
        ("provenance-input", ("data", "steps", "produce", "inputs", "history"), {"provenance": "produce"}),
    ]
    with tempfile.TemporaryDirectory(prefix="hydra-reuse-policy-") as folder:
        for name, keys, value in [("baseline", (), None)] + changes:
            plan = json.loads(json.dumps(original))
            target = plan
            for key in keys[:-1]:
                target = target[key]
            if keys:
                target[keys[-1]] = value
            path = Path(folder) / (name + ".json")
            path.write_text(json.dumps(plan))
            result = subprocess.run([str(ROOT / "bin/hydra"), "workflow", "plan", "validate",
                str(path), str(fixture / "policy.json")], cwd=source, env=env,
                capture_output=True, text=True, timeout=30)
            document = json.loads(result.stdout)
            if name == "baseline":
                assert result.returncode == 0 and document["ok"], document
            else:
                assert result.returncode != 0 and not document["ok"], (name, document)
                assert any(row["code"] == "unsupported_reuse" for row in document["data"]["diagnostics"]), (name, document)
        print("policy: baseline accepted, 9 unsupported declarations rejected")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("fixture", type=Path)
    parser.add_argument("--fleet-bin", type=Path,
                        default=Path(os.environ.get("HYDRA_FLEET_BIN", ROOT / "build/hydra-fleet")))
    args = parser.parse_args()
    fixture = args.fixture.resolve()
    original_home, run_id, original_source = find_fixture(fixture)
    original_record = json.loads((run_path(original_home, run_id) / "repair-2.json").read_text())
    assert json_digest(original_record["evidence"]) == original_record["sha256"], "canonical hash parity"
    run_result(original_source, original_home, run_id, args.fleet_bin, True)
    policy_controls(fixture, original_source, original_home, run_id, args.fleet_bin)
    cases = [("input", tamper_input), ("environment", tamper_environment),
             ("recipe", tamper_recipe),
             ("acceptance", tamper_acceptance), ("receipt-rehashed", tamper_receipt),
             ("result-rehashed", tamper_result), ("malformed-failures", tamper_failures),
             ("compiled-contract", tamper_compiled), ("source-binding", tamper_source_binding),
             ("unknown-proof-step", tamper_unknown_step), ("dependency-removal", tamper_dependency_removal)]
    with tempfile.TemporaryDirectory(prefix="hydra-plan-reuse-") as scratch:
        for name, mutate in cases:
            case = Path(scratch) / name
            shutil.copytree(original_home, case / "home")
            mutate(run_path(case / "home", run_id))
            run_result(original_source, case / "home", run_id, args.fleet_bin, False)
            print(f"{name}: rejected")
    print("baseline: accepted")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tests/native/test_plan_reuse_invalidation.c
#include "accepted_fixture.h"
static char hydra[F_PATH], source[F_PATH], fixture[F_PATH];
static const char *fleet;
static void path(char *out, const char *run, const char *name) {
  nt_path(out, run, name);
}
static json_object *read_at(const char *run, const char *name) {
  char p[F_PATH];
  path(p, run, name);
  json_object *v = f_read_json(p, 8000000);
  assert(v);
  return v;
}
static void write_at(const char *run, const char *name, json_object *v) {
  char p[F_PATH];
  path(p, run, name);
  nt_json_write(p, v);
}
static void rehash_record(const char *run, const char *name) {
  char p[F_PATH], digest[65];
  json_object *v = read_at(run, "repair-2.json"), *e = f_field(v, "evidence"),
              *proof =
                  f_field(f_field(f_field(e, "reuse"), "steps"), "produce");
  char relative[F_PATH];
  assert(snprintf(relative, sizeof relative, "steps/produce/attempt-1/%s",
                  name) > 0);
  path(p, run, relative);
  assert(!f_hash(p, digest));
  f_string_add(proof, name, digest);
  af_hash_json(e, digest, true);
  f_string_add(v, "sha256", digest);
  write_at(run, "repair-2.json", v);
  json_object_put(v);
}
static void run_result(const char *home, bool pass, bool changed_path) {
  char run[F_PATH];
  af_run_path(run, home);
  af_environment(run, home, fleet);
  if (changed_path) {
    const char *old = getenv("PATH");
    assert(old);
    char *s = malloc(strlen(old) + 40);
    assert(s);
    int n =
        snprintf(s, strlen(old) + 40, "%s:/hydra-undeclared-tool-path", old);
    assert(n > 0);
    assert(!setenv("PATH", s, 1));
    free(s);
  }
  glob_t before, after;
  af_glob(&before, home, "fleet/tasks/task_*/acceptance.json");
  const char *id = strrchr(run, '/');
  assert(id);
  char *args[] = {hydra, "workflow", "plan", "result", (char *)id + 1, NULL};
  struct f_capture c = nt_run(args);
  if (pass) {
    json_object *v = nt_output(&c, 0);
    nt_equal(f_field(v, "ok"), "true");
    json_object_put(v);
  } else {
    if (!c.status)
      fprintf(stderr, "corrupted copied fixture accepted: %s\n", c.out);
    assert(c.status);
  }
  f_capture_free(&c);
  af_glob(&after, home, "fleet/tasks/task_*/acceptance.json");
  assert(before.gl_pathc == after.gl_pathc);
  for (size_t i = 0; i < before.gl_pathc; i++)
    assert(!strcmp(before.gl_pathv[i], after.gl_pathv[i]));
  globfree(&before);
  globfree(&after);
}
static void mutation(const char *run, size_t index) {
  char p[F_PATH], digest[65];
  json_object *v, *e;
  if (index == 0 || index == 2) {
    path(p, run,
         index == 0 ? "steps/produce/attempt-1/inputs/environment"
                    : "steps/produce/attempt-1/inputs/recipe");
    FILE *f = fopen(p, "a");
    assert(f && fputc(' ', f) != EOF && !fclose(f));
    return;
  }
  if (index == 1) {
    v = read_at(run, "steps/produce/attempt-1/inputs/environment");
    f_string_add(v, "scope", "tampered");
    write_at(run, "steps/produce/attempt-1/inputs/environment", v);
    json_object_put(v);
    return;
  }
  if (index == 3) {
    path(p, run, "plan-accepted");
    nt_write(
        p,
        "0000000000000000000000000000000000000000000000000000000000000000\n");
    return;
  }
  if (index == 4) {
    v = read_at(run, "steps/produce/attempt-1/remote/receipt.json");
    f_string_add(v, "submission_key", "tampered");
    write_at(run, "steps/produce/attempt-1/remote/receipt.json", v);
    json_object_put(v);
    rehash_record(run, "remote/receipt.json");
    return;
  }
  if (index == 5 || (index >= 7 && index <= 10)) {
    v = read_at(run, "steps/produce/attempt-1/remote/result.json");
    json_object *result = f_field(v, "result"),
                *receipt = f_field(result, "receipt");
    af_hash_json(result, digest, false);
    assert(!strcmp(digest, f_string(v, "result_sha256")));
    if (index == 5)
      f_string_add(
          receipt, "task_id",
          "task_"
          "0000000000000000000000000000000000000000000000000000000000000000");
    else {
      const char *states[] = {"outcome_unknown", "failed", "cancelled",
                              "succeeded"};
      f_string_add(f_field(receipt, "runtime"), "state", states[index - 7]);
      if (index == 10)
        json_object_object_add(f_field(receipt, "runtime"), "exit_status",
                               json_object_new_int(7));
    }
    af_hash_json(result, digest, false);
    f_string_add(v, "result_sha256", digest);
    write_at(run, "steps/produce/attempt-1/remote/result.json", v);
    json_object_put(v);
    rehash_record(run, "remote/result.json");
    return;
  }
  if (index == 11) {
    path(p, run, "steps/produce/state");
    nt_write(p, "outcome_unknown\n");
    return;
  }
  if (index == 12 || index == 13) {
    v = read_at(run, "compiled.json");
    if (index == 12) {
      e = f_field(v, "plan");
      const char *old = f_string(e, "objective");
      char *s = malloc(strlen(old) + 10);
      assert(s);
      assert(snprintf(s, strlen(old) + 10, "%s tampered", old) > 0);
      f_string_add(e, "objective", s);
      free(s);
    } else
      f_string_add(
          f_field(v, "source"), "sha256",
          "0000000000000000000000000000000000000000000000000000000000000000");
    write_at(run, "compiled.json", v);
    json_object_put(v);
    return;
  }
  v = read_at(run, "repair-2.json");
  e = f_field(v, "evidence");
  if (index == 6)
    json_object_object_add(e, "failures", json_object_new_array());
  else if (index == 14)
    json_object_object_add(f_field(f_field(e, "reuse"), "steps"), "unknown",
                           json_object_new_object());
  else {
    assert(index == 15);
    json_object_object_del(f_field(f_field(e, "reuse"), "steps"), "inspect");
  }
  af_hash_json(e, digest, true);
  f_string_add(v, "sha256", digest);
  write_at(run, "repair-2.json", v);
  json_object_put(v);
}
static void policy_controls(const char *home) {
  char p[F_PATH], run[F_PATH], policy[F_PATH],
      temp[] = "/tmp/hydra-reuse-policy-XXXXXX";
  af_run_path(run, home);
  af_environment(run, home, fleet);
  path(p, fixture, "plan.json");
  json_object *original = f_read_json(p, 8000000);
  assert(original);
  path(policy, fixture, "policy.json");
  assert(mkdtemp(temp));
  path(p, temp, "plan.json");
  const char *keys[][6] = {
      {"reuse_policy", "schema_version", NULL},
      {"reuse_policy", "mode", NULL},
      {"reuse_policy", "extra", NULL},
      {"reuse_policy", "steps", "unknown", NULL},
      {"reuse_policy", "steps", "produce", "dependencies", NULL},
      {"reuse_policy", "steps", "produce", "effects", NULL},
      {"reuse_policy", "steps", "produce", "environment_input", NULL},
      {"data", "steps", "produce", "inputs", "repair", NULL},
      {"data", "steps", "produce", "inputs", "history", NULL}};
  const char *values[] = {
      "2",           "\"external_cache\"", "true",
      "{}",          "\"unknown\"",        "\"external\"",
      "\"missing\"", "{\"repair\":true}",  "{\"provenance\":\"produce\"}"};
  for (int i = -1; i < 9; i++) {
    json_object *v = nt_clone(original);
    if (i >= 0) {
      json_object *target = v;
      size_t k = 0;
      while (keys[i][k + 1]) {
        target = f_field(target, keys[i][k]);
        assert(target);
        k++;
      }
      json_object_object_add(target, keys[i][k], f_parse_value(values[i]));
    }
    nt_json_write(p, v);
    json_object_put(v);
    char *args[] = {hydra, "workflow", "plan", "validate", p, policy, NULL};
    struct f_capture c = nt_run(args);
    v = f_parse(c.out);
    assert(v);
    if (i < 0) {
      assert(!c.status);
      nt_equal(f_field(v, "ok"), "true");
    } else {
      assert(c.status);
      nt_equal(f_field(v, "ok"), "false");
      json_object *rows = f_field(f_field(v, "data"), "diagnostics");
      bool found = false;
      for (size_t j = 0; j < json_object_array_length(rows); j++)
        if (!strcmp(f_string(json_object_array_get_idx(rows, j), "code"),
                    "unsupported_reuse"))
          found = true;
      assert(found);
    }
    json_object_put(v);
    f_capture_free(&c);
  }
  json_object_put(original);
  assert(!f_remove_tree(temp));
  puts("policy: baseline accepted, 9 unsupported declarations rejected");
}
int main(int argc, char **argv) {
  char root[F_PATH], home[F_PATH], run[F_PATH], binary[F_PATH],
      scratch[] = "/tmp/hydra-plan-reuse-native-XXXXXX";
  assert(argc == 2 || argc == 4);
  assert(getcwd(root, sizeof root));
  path(hydra, root, "bin/hydra");
  if (argv[1][0] == '/')
    assert(!f_copy(fixture, sizeof fixture, argv[1]));
  else
    path(fixture, root, argv[1]);
  fleet = getenv("HYDRA_FLEET_BIN");
  if (argc == 4) {
    assert(!strcmp(argv[2], "--fleet-bin"));
    fleet = argv[3];
  }
  if (!fleet) {
    path(binary, root, "build/hydra-fleet");
    fleet = binary;
  }
  path(home, fixture, "home");
  path(source, fixture, "source");
  af_run_path(run, home);
  json_object *v = read_at(run, "repair-2.json");
  char digest[65];
  af_hash_json(f_field(v, "evidence"), digest, true);
  assert(!strcmp(digest, f_string(v, "sha256")));
  json_object_put(v);
  assert(!chdir(source));
  run_result(home, true, false);
  run_result(home, false, true);
  puts("changed-current-environment: rejected without new acceptance");
  policy_controls(home);
  assert(mkdtemp(scratch));
  const char *names[] = {"input",
                         "environment",
                         "recipe",
                         "acceptance",
                         "receipt-rehashed",
                         "result-rehashed",
                         "malformed-failures",
                         "unknown-result-rehashed",
                         "failed-result-rehashed",
                         "cancelled-result-rehashed",
                         "nonzero-success-result",
                         "unknown-step-outcome",
                         "compiled-contract",
                         "source-binding",
                         "unknown-proof-step",
                         "dependency-removal"};
  for (size_t i = 0; i < 16; i++) {
    char copy[F_PATH];
    path(copy, scratch, names[i]);
    nt_copy_tree(home, copy);
    af_run_path(run, copy);
    mutation(run, i);
    run_result(copy, false, false);
    printf("%s: rejected\n", names[i]);
  }
  assert(!chdir(root));
  nt_finish(scratch, "baseline: accepted");
  return 0;
}
